Add support for Web Extension browser.tabs events.

xeenon opened the pull request:
#### 2c6a9a12e9abd0dec08fc0dfaad2c27493b62b7a
<pre>
Add support for Web Extension browser.tabs events.
<a href="https://webkit.org/b/261681">https://webkit.org/b/261681</a>
rdar://problem/115663241

Reviewed by Brian Weinstein.

* Implement all the tabs events that are supported by Safari.
* Internalize and share more of the windows and tabs test code in TestWebKitAPI so
  individual tests are simplier and don&apos;t need to repeat the same stuff.
* Flip the actual and expected arguments on test.assertEq and assertDeepEq to
  match the actual usage in the tests and make logging bad results make sense.

* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionContext.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionContext.mm:
(-[_WKWebExtensionContext didActivateTab:previousActiveTab:]): Added previousActiveTab param.
(-[_WKWebExtensionContext didSelectTabs:]): Call renamed didSelectOrDeselectTabs method.
(-[_WKWebExtensionContext didDeselectTabs:]): Added. Call didSelectOrDeselectTabs method.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionController.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionController.mm:
(-[_WKWebExtensionController didActivateTab:previousActiveTab:]): Added previousActiveTab param.
(-[_WKWebExtensionController didDeselectTabs:]): Added.
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIAlarmsCocoa.mm:
(WebKit::WebExtensionContext::fireAlarmsEventIfNeeded): Use constexpr.
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPITabsCocoa.mm:
(WebKit::WebExtensionContext::fireTabsCreatedEventIfNeeded): Added.
(WebKit::WebExtensionContext::fireTabsUpdatedEventIfNeeded): Added.
(WebKit::WebExtensionContext::fireTabsReplacedEventIfNeeded): Added.
(WebKit::WebExtensionContext::fireTabsDetachedEventIfNeeded): Added.
(WebKit::WebExtensionContext::fireTabsMovedEventIfNeeded): Added.
(WebKit::WebExtensionContext::fireTabsAttachedEventIfNeeded): Added.
(WebKit::WebExtensionContext::fireTabsActivatedEventIfNeeded): Added.
(WebKit::WebExtensionContext::fireTabsHighlightedEventIfNeeded): Added.
(WebKit::WebExtensionContext::fireTabsRemovedEventIfNeeded): Added.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::didOpenTab): Fire the event.
(WebKit::WebExtensionContext::didCloseTab): Fire the event.
(WebKit::WebExtensionContext::didActivateTab): Fire the event.
(WebKit::WebExtensionContext::didSelectOrDeselectTabs): Fire the event.
(WebKit::WebExtensionContext::didMoveTab): Fire the event.
(WebKit::WebExtensionContext::didReplaceTab): Fire the event.
(WebKit::WebExtensionContext::didChangeTabProperties): Fire the event.
(WebKit::WebExtensionContext::didSelectTabs): Deleted. Renamed to didSelectOrDeselectTabs.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionTabCocoa.mm:
(WebKit::WebExtensionTab::parameters const):
(WebKit::WebExtensionTab::changedParameters const): Added.
(WebKit::WebExtensionTab::window const):
(WebKit::WebExtensionTab::minimalParameters const): Deleted.
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionTab.h:
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPITabsCocoa.mm:
(WebKit::toWebAPI): Only incldue properties that have values so changedParameters only includes what changed.
(WebKit::WebExtensionContextProxy::dispatchTabsCreatedEvent): Added.
(WebKit::WebExtensionContextProxy::dispatchTabsUpdatedEvent): Added.
(WebKit::WebExtensionContextProxy::dispatchTabsReplacedEvent): Added.
(WebKit::WebExtensionContextProxy::dispatchTabsDetachedEvent): Added.
(WebKit::WebExtensionContextProxy::dispatchTabsMovedEvent): Added.
(WebKit::WebExtensionContextProxy::dispatchTabsAttachedEvent): Added.
(WebKit::WebExtensionContextProxy::dispatchTabsActivatedEvent): Added.
(WebKit::WebExtensionContextProxy::dispatchTabsHighlightedEvent): Added.
(WebKit::WebExtensionContextProxy::dispatchTabsRemovedEvent): Added.
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPITestCocoa.mm:
(WebKit::WebExtensionAPITest::assertDeepEq): Flipped arguments.
(WebKit::WebExtensionAPITest::assertEq): Flipped arguments.
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIWindowsCocoa.mm:
(WebKit::WebExtensionAPIWindows::parseWindowUpdateOptions): Added an exception for bad state.
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPITest.h: Flipped arguments.
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPITest.idl: Flipped arguments.
* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.h:
* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.messages.in: Added tab events.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPITabs.mm:
(TestWebKitAPI::TEST): Added new event tests and simplified existing tests.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIWindows.mm:
(TestWebKitAPI::TEST): Simplified.
* Tools/TestWebKitAPI/cocoa/TestWebExtensionsDelegate.mm:
(-[TestWebExtensionsDelegate webExtensionController:openNewTabWithOptions:forExtensionContext:completionHandler:]):
Added default window opening logic.
* Tools/TestWebKitAPI/cocoa/WebExtensionUtilities.h:
* Tools/TestWebKitAPI/cocoa/WebExtensionUtilities.mm:
(-[TestWebExtensionManager initForExtension:]): Added default block implementations.
(-[TestWebExtensionManager openNewWindow]): Added.
(-[TestWebExtensionManager focusWindow:]): Added.
(-[TestWebExtensionManager closeWindow:]): Added.
(-[TestWebExtensionManager run]): Added whitespace.
(-[TestWebExtensionTab init]): Added to keep other tests working.
(-[TestWebExtensionTab initWithWindow:extensionController:]): Don&apos;t make web view is no extension controller,
this is needed for some tests still and it avoids an exception.
(-[TestWebExtensionTab toggleReaderModeForWebExtensionContext:completionHandler:]): Call didChangeTabProperties:.
(-[TestWebExtensionTab reloadForWebExtensionContext:completionHandler:]): Added default implementation.
(-[TestWebExtensionTab reloadFromOriginForWebExtensionContext:completionHandler:]): Added default implementation.
(-[TestWebExtensionTab goBackForWebExtensionContext:completionHandler:]): Added default implementation.
(-[TestWebExtensionTab goForwardForWebExtensionContext:completionHandler:]): Added default implementation.
(-[TestWebExtensionTab pinForWebExtensionContext:completionHandler:]): Call didChangeTabProperties:.
(-[TestWebExtensionTab unpinForWebExtensionContext:completionHandler:]): Call didChangeTabProperties:.
(-[TestWebExtensionTab muteForWebExtensionContext:completionHandler:]): Call didChangeTabProperties:.
(-[TestWebExtensionTab unmuteForWebExtensionContext:completionHandler:]): Call didChangeTabProperties:.
(-[TestWebExtensionTab activateForWebExtensionContext:completionHandler:]): Call didActivateTab:.
(-[TestWebExtensionTab selectForWebExtensionContext:completionHandler:]): Call didSelectTabs:.
(-[TestWebExtensionTab deselectForWebExtensionContext:completionHandler:]): Call didDeselectTabs:.
(-[TestWebExtensionWindow init]): Call initWithExtensionController:.
(-[TestWebExtensionWindow initWithExtensionController:]): Added.
(-[TestWebExtensionWindow openNewTab]): Added.
(-[TestWebExtensionWindow openNewTabAtIndex:]): Added.
(-[TestWebExtensionWindow closeTab:]): Added.
(-[TestWebExtensionWindow replaceTab:withTab:]): Added.
(-[TestWebExtensionWindow moveTab:toIndex:]): Added.

Canonical link: <a href="https://commits.webkit.org/268091@main">https://commits.webkit.org/268091@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a188f455f64df545641e580f95388d99ed4b9707

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/18662 "3 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/19003 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/19606 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/20525 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/17469 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/18858 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/22312 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/19144 "Built successfully") | [⏳ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/WPE-WK2-Tests-EWS "Waiting to run tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/18887 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/19044 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/16249 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/21402 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/16267 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/17009 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/23460 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/17292 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/17181 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/21353 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/17783 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/15094 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/16836 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/21203 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/2286 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/17622 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

b-weinstein commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -1427,53 +1427,90 @@ static _WKWebExtensionContextError toAPI(WebExtensionContext::Error error)
+void WebExtensionContext::didActivateTab(const WebExtensionTab& tab, const WebExtensionTab* previousTab)
 {
     ASSERT(tab.extensionContext() == this);
+    ASSERT(!previousTab || previousTab && previousTab->extensionContext() == this);
```
Should the second part of the ASSERT have parens? Otherwise, I'm not sure how useful:

`ASSERT(!previousTab || previousTab` is.

`ASSERT(!previousTab || (previousTab && previousTab->extensionContext() == this));`

b-weinstein commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPITabsCocoa.mm:
```diff
@@ -846,6 +850,108 @@ static bool isValid(std::optional<WebExtensionTabIdentifier> identifier, NSStrin
+{
+    // Documentation: https://developer.mozilla.org/docs/Mozilla/Add-ons/WebExtensions/API/tabs/onActivated
+
+    auto *activeInfo = @{ previousTabIdKey: @(toWebAPI(previousActiveTabIdentifier)), tabIdKey: @(toWebAPI(newActiveTabIdentifier)), windowIdKey: @(toWebAPI(windowIdentifier)) };
```
Super-nit: `activateInfo` might be more clear.

xeenon committed:
Add support for Web Extension browser.tabs events.
https://webkit.org/b/261681
rdar://problem/115663241

Reviewed by Brian Weinstein.

* Implement all the tabs events that are supported by Safari.
* Internalize and share more of the windows and tabs test code in TestWebKitAPI so
  individual tests are simplier and don't need to repeat the same stuff.
* Flip the actual and expected arguments on test.assertEq and assertDeepEq to
  match the actual usage in the tests and make logging bad results make sense.

* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionContext.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionContext.mm:
(-[_WKWebExtensionContext didActivateTab:previousActiveTab:]): Added previousActiveTab param.
(-[_WKWebExtensionContext didSelectTabs:]): Call renamed didSelectOrDeselectTabs method.
(-[_WKWebExtensionContext didDeselectTabs:]): Added. Call didSelectOrDeselectTabs method.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionController.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionController.mm:
(-[_WKWebExtensionController didActivateTab:previousActiveTab:]): Added previousActiveTab param.
(-[_WKWebExtensionController didDeselectTabs:]): Added.
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIAlarmsCocoa.mm:
(WebKit::WebExtensionContext::fireAlarmsEventIfNeeded): Use constexpr.
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPITabsCocoa.mm:
(WebKit::WebExtensionContext::fireTabsCreatedEventIfNeeded): Added.
(WebKit::WebExtensionContext::fireTabsUpdatedEventIfNeeded): Added.
(WebKit::WebExtensionContext::fireTabsReplacedEventIfNeeded): Added.
(WebKit::WebExtensionContext::fireTabsDetachedEventIfNeeded): Added.
(WebKit::WebExtensionContext::fireTabsMovedEventIfNeeded): Added.
(WebKit::WebExtensionContext::fireTabsAttachedEventIfNeeded): Added.
(WebKit::WebExtensionContext::fireTabsActivatedEventIfNeeded): Added.
(WebKit::WebExtensionContext::fireTabsHighlightedEventIfNeeded): Added.
(WebKit::WebExtensionContext::fireTabsRemovedEventIfNeeded): Added.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::didOpenTab): Fire the event.
(WebKit::WebExtensionContext::didCloseTab): Fire the event.
(WebKit::WebExtensionContext::didActivateTab): Fire the event.
(WebKit::WebExtensionContext::didSelectOrDeselectTabs): Fire the event.
(WebKit::WebExtensionContext::didMoveTab): Fire the event.
(WebKit::WebExtensionContext::didReplaceTab): Fire the event.
(WebKit::WebExtensionContext::didChangeTabProperties): Fire the event.
(WebKit::WebExtensionContext::didSelectTabs): Deleted. Renamed to didSelectOrDeselectTabs.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionTabCocoa.mm:
(WebKit::WebExtensionTab::parameters const):
(WebKit::WebExtensionTab::changedParameters const): Added.
(WebKit::WebExtensionTab::window const):
(WebKit::WebExtensionTab::minimalParameters const): Deleted.
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionTab.h:
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPITabsCocoa.mm:
(WebKit::toWebAPI): Only incldue properties that have values so changedParameters only includes what changed.
(WebKit::WebExtensionContextProxy::dispatchTabsCreatedEvent): Added.
(WebKit::WebExtensionContextProxy::dispatchTabsUpdatedEvent): Added.
(WebKit::WebExtensionContextProxy::dispatchTabsReplacedEvent): Added.
(WebKit::WebExtensionContextProxy::dispatchTabsDetachedEvent): Added.
(WebKit::WebExtensionContextProxy::dispatchTabsMovedEvent): Added.
(WebKit::WebExtensionContextProxy::dispatchTabsAttachedEvent): Added.
(WebKit::WebExtensionContextProxy::dispatchTabsActivatedEvent): Added.
(WebKit::WebExtensionContextProxy::dispatchTabsHighlightedEvent): Added.
(WebKit::WebExtensionContextProxy::dispatchTabsRemovedEvent): Added.
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPITestCocoa.mm:
(WebKit::WebExtensionAPITest::assertDeepEq): Flipped arguments.
(WebKit::WebExtensionAPITest::assertEq): Flipped arguments.
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIWindowsCocoa.mm:
(WebKit::WebExtensionAPIWindows::parseWindowUpdateOptions): Added an exception for bad state.
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPITest.h: Flipped arguments.
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPITest.idl: Flipped arguments.
* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.h:
* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.messages.in: Added tab events.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPITabs.mm:
(TestWebKitAPI::TEST): Added new event tests and simplified existing tests.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIWindows.mm:
(TestWebKitAPI::TEST): Simplified.
* Tools/TestWebKitAPI/cocoa/TestWebExtensionsDelegate.mm:
(-[TestWebExtensionsDelegate webExtensionController:openNewTabWithOptions:forExtensionContext:completionHandler:]):
Added default window opening logic.
* Tools/TestWebKitAPI/cocoa/WebExtensionUtilities.h:
* Tools/TestWebKitAPI/cocoa/WebExtensionUtilities.mm:
(-[TestWebExtensionManager initForExtension:]): Added default block implementations.
(-[TestWebExtensionManager openNewWindow]): Added.
(-[TestWebExtensionManager focusWindow:]): Added.
(-[TestWebExtensionManager closeWindow:]): Added.
(-[TestWebExtensionManager run]): Added whitespace.
(-[TestWebExtensionTab init]): Added to keep other tests working.
(-[TestWebExtensionTab initWithWindow:extensionController:]): Don't make web view is no extension controller,
this is needed for some tests still and it avoids an exception.
(-[TestWebExtensionTab toggleReaderModeForWebExtensionContext:completionHandler:]): Call didChangeTabProperties:.
(-[TestWebExtensionTab reloadForWebExtensionContext:completionHandler:]): Added default implementation.
(-[TestWebExtensionTab reloadFromOriginForWebExtensionContext:completionHandler:]): Added default implementation.
(-[TestWebExtensionTab goBackForWebExtensionContext:completionHandler:]): Added default implementation.
(-[TestWebExtensionTab goForwardForWebExtensionContext:completionHandler:]): Added default implementation.
(-[TestWebExtensionTab pinForWebExtensionContext:completionHandler:]): Call didChangeTabProperties:.
(-[TestWebExtensionTab unpinForWebExtensionContext:completionHandler:]): Call didChangeTabProperties:.
(-[TestWebExtensionTab muteForWebExtensionContext:completionHandler:]): Call didChangeTabProperties:.
(-[TestWebExtensionTab unmuteForWebExtensionContext:completionHandler:]): Call didChangeTabProperties:.
(-[TestWebExtensionTab activateForWebExtensionContext:completionHandler:]): Call didActivateTab:.
(-[TestWebExtensionTab selectForWebExtensionContext:completionHandler:]): Call didSelectTabs:.
(-[TestWebExtensionTab deselectForWebExtensionContext:completionHandler:]): Call didDeselectTabs:.
(-[TestWebExtensionWindow init]): Call initWithExtensionController:.
(-[TestWebExtensionWindow initWithExtensionController:]): Added.
(-[TestWebExtensionWindow openNewTab]): Added.
(-[TestWebExtensionWindow openNewTabAtIndex:]): Added.
(-[TestWebExtensionWindow closeTab:]): Added.
(-[TestWebExtensionWindow replaceTab:withTab:]): Added.
(-[TestWebExtensionWindow moveTab:toIndex:]): Added.

Canonical link: https://commits.webkit.org/268091@main
